File: CSharp/11-ContainerWithMostWater/11ContainerWithMostWater.cs
// Two Pointers

// Time Complexity: O(n)
// we have to visit each n elements of height once.

// Space Complexity: O(1)
// We are just traversing the height array that was passed in so no extra memory needed.

public class Solution {
    public int MaxArea(int[] height) {
        int maxArea = 0;
        int left = 0, right = height.Length - 1;
        while (left < right){
            int area = (right - left) * Math.Min(height[left], height[right]);
            maxArea = Math.Max(maxArea, area);
            if(height[left] < height[right]) left++;
            else right--;
        }
        return maxArea;
    }
}

File: Javascript/11-ContainerWithMostWater/11ContainerWithMostWater.js
/*

Time Complexity: O(n)
The two pointers may have to traverse every element in the array to find the highest area.

Space Complexity: O(1)
We're just iterating over the height number array with a set number of pointers.

*/

/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  let result = 0,
    left = 0,
    right = height.length - 1;
  while (left < right) {
    let smallestSide = Math.min(height[left], height[right]);
    let area = (right - left) * smallestSide;
    if (area > result) result = area;
    height[left] < height[right] ? left++ : right--;
  }
  return result;
};
